### PR TITLE
Add an optional Far Edge test case to ensure limited use of exec probes

### PR DESCRIPTION
**Pull request by shimritproj**
The entire cluster's worth of pods needs to have less than or equal to 10 exec probes.

### CODE REVIEW

File: CATALOG.md [hunk 1]
```diff
@@ -41,18 +41,34 @@ Depending on the CNF type, not all tests are required to pass to satisfy best pr
...
+#### access-control-
+
+Property|Description
+---|---
+Unique ID|access-control-
+Description|C
+Suggested Remediation|Ensure that pods use projected volumes and service account tokens
+Best Practice Reference|No Doc Link - Far Edge
+Exception Process|E.
+Tags|telco,access-control
```
Comment (ramperher) [inline CATALOG.md:65]: if the test case is for far-edge, shouldn't this be located in faredge rather than in telco tests?

**Review by ramperher — COMMENTED** [review]
The DCI job is failing because this new test is failing, but if the test is expected for far-edge, then the test should be omitted because we're not considering far-edge in our DCI jobs

File: cnf-certification-test/identifiers/identifiers.go [hunk 2]
```diff
@@ -213,6 +214,21 @@ func InitCatalog() map[claim.Identifier]claim.TestCaseDescription {
 		},
 		TagExtended)
 
+	TestLimitedUseOfExecProbesIdentifier = AddCatalogEntry(
+		"test case to ensure limited use of exec probes",
+		common.AccessControlTestKey,
```
Comment (edcdavid) [inline cnf-certification-test/identifiers/identifiers.go:219]: should be common.PerformanceTestKey

File: cnf-certification-test/identifiers/identifiers.go [hunk 3]
```diff
@@ -213,6 +214,21 @@ func InitCatalog() map[claim.Identifier]claim.TestCaseDescription {
...
+		LimitedUseOfExecProbesRemediation,
+		`Ensure that limited use of exec probes`,
+		TestLimitedUseOfExecProbesIdentifierDocLink,
+		true,
+		map[string]string{
+			FarEdge:  Optional,
+			Telco:    Optional,
+			NonTelco: Optional,
+			Extended: Optional},
+		TagTelco)
```
Comment (edcdavid) [inline cnf-certification-test/identifiers/identifiers.go:230]: TagFarEdge?

File: cnf-certification-test/performance/suite.go [hunk 4]
```diff
@@ -79,6 +85,22 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
 	// Scheduling related tests ends here
 })
 
+func testLimitedUseOfExecProbes(env *provider.TestEnvironment) {
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.LivenessProbe.PeriodSeconds < 10 && cut.LivenessProbe.InitialDelaySeconds <= 10 {
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:93]: This is about limiting the number and frequency of exec probes. Readiness, liveliness and startup probes can all  be exec probes. Consider using existing function func (c *Container) HasExecProbes() bool to only count exec probes. 
Why test InitialDelaySeconds? 
Also, there is a missing test. There should be no more that 10 exec probes (of any type) in the cluster.

File: cnf-certification-test/performance/suite.go [hunk 5]
```diff
@@ -79,6 +85,22 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+				nonCompliantObjects = append(nonCompliantObjects, testhelper.NewPodReportObject(put.Namespace, put.Name, "Ensure limited use of exec probes", false))
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:94]: The report object should match the object being tested. In this case, you should report containers. Adding the content of the exec probe would also be helpful.

File: cnf-certification-test/performance/suite.go [hunk 6]
```diff
@@ -79,6 +85,22 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+				continue
+			}
+			tnf.ClaimFilePrintf("Pod %s/Container %s defines not ensure limited use of exec probes", cut.Podname, cut)
+			compliantObjects = append(compliantObjects, testhelper.NewPodReportObject(put.Namespace, put.Name, "", true))
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:98]: same here

File: cnf-certification-test/identifiers/identifiers.go (hunk 3 above)
Comment (shimritproj) [reply cnf-certification-test/identifiers/identifiers.go:230]: You're right!

File: CATALOG.md (hunk 1 above)
Comment (shimritproj) [reply CATALOG.md]: yes , my mistake.

File: cnf-certification-test/performance/suite.go [hunk 7]
```diff
@@ -79,6 +89,32 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+	for _, put := range env.Pods {
+		counter := 0
+		for _, cut := range put.Containers {
+			if cut.HasExecProbes() {
+				tnf.ClaimFilePrintf("Pod %s/Container %s use of exec probes, InitialDelaySeconds of LivenessProbe: %s,  InitialDelaySeconds of ReadinessProbe: %s, InitialDelaySeconds of StartupProbe: %s", cut.Podname, cut,
+					cut.LivenessProbe.InitialDelaySeconds, cut.ReadinessProbe.InitialDelaySeconds, cut.StartupProbe.InitialDelaySeconds)
+				counter++
+				continue
+			}
+			tnf.ClaimFilePrintf("Pod %s/Container %s not use of exec probes, InitialDelaySeconds of LivenessProbe: %s,  InitialDelaySeconds of ReadinessProbe: %s, InitialDelaySeconds of StartupProbe: %s", cut.Podname, cut,
```
Comment (sebrandon1) [inline cnf-certification-test/performance/suite.go:104]:
```suggestion
			tnf.ClaimFilePrintf("Pod %s/Container %s is not using exec probes, InitialDelaySeconds of LivenessProbe: %s,  InitialDelaySeconds of ReadinessProbe: %s, InitialDelaySeconds of StartupProbe: %s", cut.Podname, cut,
```

File: cnf-certification-test/performance/suite.go [hunk 8]
```diff
@@ -79,6 +89,32 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+func testLimitedUseOfExecProbes(env *provider.TestEnvironment) {
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
+	for _, put := range env.Pods {
+		counter := 0
+		for _, cut := range put.Containers {
+			if cut.HasExecProbes() {
+				tnf.ClaimFilePrintf("Pod %s/Container %s use of exec probes, InitialDelaySeconds of LivenessProbe: %s,  InitialDelaySeconds of ReadinessProbe: %s, InitialDelaySeconds of StartupProbe: %s", cut.Podname, cut,
+					cut.LivenessProbe.InitialDelaySeconds, cut.ReadinessProbe.InitialDelaySeconds, cut.StartupProbe.InitialDelaySeconds)
+				counter++
```
Comment (sebrandon1) [inline cnf-certification-test/performance/suite.go:101]: If the container has `Exec` probes, check each `Liveness`, `Readiness`, and `Startup` probes for their PeriodSeconds <= 10.  If their `Exec` probe has a PeriodSeconds > 10, it's a fail.

Also, we need to count the number of `Exec` probes in the entire cluster, so just move the counter outside of the Pods loop.  There cannot be more than 10 `Exec` probes in the entire cluster.

File: cnf-certification-test/performance/suite.go [hunk 9]
```diff
@@ -79,6 +89,32 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+				continue
+			}
+			tnf.ClaimFilePrintf("Pod %s/Container %s not use of exec probes, InitialDelaySeconds of LivenessProbe: %s,  InitialDelaySeconds of ReadinessProbe: %s, InitialDelaySeconds of StartupProbe: %s", cut.Podname, cut,
+				cut.LivenessProbe.InitialDelaySeconds, cut.ReadinessProbe.InitialDelaySeconds, cut.StartupProbe.InitialDelaySeconds)
+		}
+		if counter <= LimitedUseOfExecProbes {
```
Comment (sebrandon1) [inline cnf-certification-test/performance/suite.go:107]:
```suggestion
		if counter > LimitedUseOfExecProbes {
```

**Review by sebrandon1 — CHANGES_REQUESTED** [review]
Some suggestions!

File: cnf-certification-test/performance/suite.go [hunk 10]
```diff
@@ -33,6 +33,10 @@ import (
 	"github.com/test-network-function/cnf-certification-test/pkg/tnf"
 )
 
+const (
+	LimitedUseOfExecProbes = 10
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:37]: add separate parameters for minimum period in seconds and maximum number of exec probes per CNF, LimitedUseOfExecProbes seems to cover both

File: cnf-certification-test/performance/suite.go [hunk 11]
```diff
@@ -79,6 +89,75 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.HasExecProbes() {
+				if cut.LivenessProbe.Exec != nil {
+					counter++
+					if cut.LivenessProbe.PeriodSeconds <= LimitedUseOfExecProbes {
+						compliantObjects = appendToCompliantObject(compliantObjects, cut, put)
+						counter++
+					} else {
+						// Exec probes are BAD
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:120]: remove log

File: cnf-certification-test/performance/suite.go [hunk 12]
```diff
@@ -79,6 +89,75 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+		}
+	}
+
+	// If there >10 exec probes, mark the entire cluster as a failure
+	if counter >= LimitedUseOfExecProbes {
+		tnf.ClaimFilePrintf("CNF has 10 or more exec probes")
+		nonCompliantObjects = append(nonCompliantObjects, testhelper.NewReportObject("CNF has 10 or more exec probes", testhelper.CustomResourceDefinitionType, false))
+	} else {
+		// Compliant cluster
+		compliantObjects = append(compliantObjects, testhelper.NewReportObject("CNF has less than 10 exec probes", testhelper.CustomResourceDefinitionType, true))
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:154]: same here

File: cnf-certification-test/performance/suite.go [hunk 13]
```diff
@@ -79,6 +89,75 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+			}
+		}
+	}
+
+	// If there >10 exec probes, mark the entire cluster as a failure
+	if counter >= LimitedUseOfExecProbes {
+		tnf.ClaimFilePrintf("CNF has 10 or more exec probes")
+		nonCompliantObjects = append(nonCompliantObjects, testhelper.NewReportObject("CNF has 10 or more exec probes", testhelper.CustomResourceDefinitionType, false))
+	} else {
+		// Compliant cluster
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:153]: // Compliant cluster -> // Compliant objects

**Comment by edcdavid** [comment]
some more comments

File: cnf-certification-test/performance/suite.go [hunk 14]
```diff
@@ -79,6 +89,75 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+	// The entire cluster's worth of pods needs to have less than or equal to 10 exec probes.
+	counter := 0
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.HasExecProbes() {
+				if cut.LivenessProbe.Exec != nil {
+					counter++
+					if cut.LivenessProbe.PeriodSeconds <= LimitedUseOfExecProbes {
+						compliantObjects = appendToCompliantObject(compliantObjects, cut, put)
+						counter++
```
[the remark this answers is not in the document]
Comment (shimritproj) [reply cnf-certification-test/performance/suite.go:118]: You're right.

File: cnf-certification-test/identifiers/identifiers.go [hunk 15]
```diff
@@ -213,6 +214,21 @@ func InitCatalog() map[claim.Identifier]claim.TestCaseDescription {
 		},
 		TagExtended)
 
+	TestLimitedUseOfExecProbesIdentifier = AddCatalogEntry(
+		"test case to ensure limited use of exec probes",
```
Comment (edcdavid) [inline cnf-certification-test/identifiers/identifiers.go:218]: Fix Format the test case name, like: xxx-yyy-aaa. For example: max-exec-probes

File: cnf-certification-test/identifiers/identifiers.go [hunk 16]
```diff
@@ -213,6 +214,21 @@ func InitCatalog() map[claim.Identifier]claim.TestCaseDescription {
...
+		common.PerformanceTestKey,
+		`The entire cluster's worth of pods needs to have less than or equal to 10 exec probes.`,
```
Comment (edcdavid) [inline cnf-certification-test/identifiers/identifiers.go:220]: add the test on the period in the description

**Review by ramperher — APPROVED** [review]
LGTM from DCI side

**Review by edcdavid — COMMENTED** [review]
adding some pending comments that I forgot to push

File: cnf-certification-test/performance/suite.go [hunk 17]
```diff
@@ -79,6 +90,69 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
 
+func testLimitedUseOfExecProbes(env *provider.TestEnvironment) {
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
+	counter := 0
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.LivenessProbe != nil && cut.LivenessProbe.Exec != nil {
+				counter++
+				if cut.LivenessProbe.PeriodSeconds <= MaximumOfExecProbes && cut.LivenessProbe.PeriodSeconds > MinimumOfPeriod {
```
Comment (aabughosh) [inline cnf-certification-test/performance/suite.go:101]: i think u dont need cut.LivenessProbe.PeriodSeconds > MinimumOfPeriod
u are checking if somthing <=10 && somthing >10

File: cnf-certification-test/performance/suite.go [hunk 18]
```diff
@@ -79,6 +90,69 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+						cut.LivenessProbe.PeriodSeconds, cut.ReadinessProbe.PeriodSeconds, cut.StartupProbe.PeriodSeconds)
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID,
+						"Ensure limited is use of exec probes", false))
+				}
+			}
+		}
+	}
+
+	// If there >10 exec probes, mark the entire cluster as a failure
+	if counter >= MaximumOfExecProbes && counter < MinimumOfPeriod {
```
[the remark this answers is not in the document]
Comment (shimritproj) [reply cnf-certification-test/performance/suite.go]: I agree with you that the condition need to be counter>=MaximumOfExecProbes, my mistake.

File: cnf-certification-test/performance/suite.go [hunk 19]
```diff
@@ -79,6 +90,69 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
 
+func testLimitedUseOfExecProbes(env *provider.TestEnvironment) {
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
+	counter := 0
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.LivenessProbe != nil && cut.LivenessProbe.Exec != nil {
+				counter++
+				if cut.LivenessProbe.PeriodSeconds <= maxNumberOfExecProbes && cut.LivenessProbe.PeriodSeconds > minExecProbePeriod {
```
Comment (sebrandon1) [inline cnf-certification-test/performance/suite.go:101]:
```suggestion
				if cut.LivenessProbe.PeriodSeconds > minExecProbePeriod {
```

I think you only need to check this condition.  Same with the rest of the following similar sections of code.

File: cnf-certification-test/performance/suite.go [hunk 20]
```diff
@@ -79,6 +90,69 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+					tnf.ClaimFilePrintf("Container %s is not using of exec probes, PeriodSeconds of ReadinessProbe: %s", cut,
+						cut.ReadinessProbe.PeriodSeconds)
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID,
+						"Ensure limited is use of exec probes", false))
+				}
+			}
+		}
+	}
+
+	// If there >10 exec probes, mark the entire cluster as a failure
```
Comment (sebrandon1) [inline cnf-certification-test/performance/suite.go:141]:
```suggestion
	// If there >= 10 exec probes, mark the entire cluster as a failure
```

**Review by sebrandon1 — COMMENTED** [review]
Some more comments!  I think this is close just needs a few more tweaks.

File: cnf-certification-test/performance/suite.go (hunk 19 above)
Comment (aabughosh) [reply cnf-certification-test/performance/suite.go:101]: agree with brandon

Comment (shimritproj) [reply cnf-certification-test/performance/suite.go:101]: I agree

File: cnf-certification-test/identifiers/identifiers.go [hunk 21]
```diff
@@ -213,6 +214,21 @@ func InitCatalog() map[claim.Identifier]claim.TestCaseDescription {
 		},
 		TagExtended)
 
+	TestLimitedUseOfExecProbesIdentifier = AddCatalogEntry(
+		"max-resources-exec-probes",
+		common.PerformanceTestKey,
+		`The entire cluster's worth of pods needs to have less than or equal to 10 exec probes and ensure that resources using the field periodSeconds is greater than 10`,
```
Comment (edcdavid) [inline cnf-certification-test/identifiers/identifiers.go:220]: "The entire cluster's worth of pods needs to have less than or equal to 10 exec probes and ensure that resources using the field periodSeconds is greater than 10." -> "Checks that 10 or less exec probes are configured in the cluster for this CNF. Also checks that the periodSeconds parameter for each probe is superior or equal to 10."

File: cnf-certification-test/identifiers/remediation.go [hunk 22]
```diff
@@ -152,6 +152,8 @@ const (
 
 	ProjectedVolumeServiceAccountRemediation = `Ensure that pods do not use projected volumes and service account tokens`
 
+	LimitedUseOfExecProbesRemediation = `The entire cluster's worth of pods needs to have less than or equal to 10 exec probes.`
```
Comment (edcdavid) [inline cnf-certification-test/identifiers/remediation.go:155]: "The entire cluster's worth of pods needs to have less than or equal to 10 exec probes." -> "Reduce the number of exec probes in the cluster for this CNF to less than 10. Increase the update period of the exec probe to be superior or equal to 10 seconds.

File: cnf-certification-test/performance/suite.go [hunk 23]
```diff
@@ -33,6 +33,11 @@ import (
 	"github.com/test-network-function/cnf-certification-test/pkg/tnf"
 )
 
+const (
+	maxNumberOfExecProbes = 10
+	minExecProbePeriod    = 10
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:38]: minExecProbePeriod -> minExecProbePeriodSeconds

File: cnf-certification-test/performance/suite.go [hunk 24]
```diff
@@ -79,6 +90,66 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+			if cut.LivenessProbe != nil && cut.LivenessProbe.Exec != nil {
+				counter++
+				if cut.LivenessProbe.PeriodSeconds > minExecProbePeriod {
+					tnf.ClaimFilePrintf("Container %s is using exec probes, PeriodSeconds of LivenessProbe: %s", cut,
+						cut.LivenessProbe.PeriodSeconds)
+					compliantObjects = append(compliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID, "Liveness exec probe has a PeriodSeconds greater than 10", true))
+				} else {
+					tnf.ClaimFilePrintf("Container %s is not using of exec probes, PeriodSeconds of LivenessProbe: %s", cut,
+						cut.LivenessProbe.PeriodSeconds)
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID, "Ensure limited use of exec probes", false))
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:108]: "Ensure limited use of exec probes" - > "Liveness exec probe has a PeriodSeconds less than 10."
Same for startup and Readiness probes

File: cnf-certification-test/performance/suite.go [hunk 25]
```diff
@@ -79,6 +90,66 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+					tnf.ClaimFilePrintf("Container %s is not using of exec probes, PeriodSeconds of ReadinessProbe: %s", cut,
+						cut.ReadinessProbe.PeriodSeconds)
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID, "Ensure limited use of exec probes", false))
+				}
+			}
+		}
+	}
+
+	// If there >=10 exec probes, mark the entire cluster as a failure
+	if counter >= maxNumberOfExecProbes {
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:139]: >= should be >. 10 is ok

File: cnf-certification-test/performance/suite.go [hunk 26]
```diff
@@ -79,6 +90,66 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+		tnf.ClaimFilePrintf("CNF has 10 or more exec probes")
+		nonCompliantObjects = append(nonCompliantObjects, testhelper.NewReportObject("CNF has 10 or more exec probes", testhelper.CnfType, false).
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:141]: "CNF has 10 or more exec probes" -> fmt.Sprintf("CNF has %d exec probes",counter)

File: cnf-certification-test/performance/suite.go [hunk 27]
```diff
@@ -79,6 +90,66 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+	var nonCompliantObjects []*testhelper.ReportObject
+	counter := 0
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.LivenessProbe != nil && cut.LivenessProbe.Exec != nil {
+				counter++
+				if cut.LivenessProbe.PeriodSeconds > minExecProbePeriod {
+					tnf.ClaimFilePrintf("Container %s is using exec probes, PeriodSeconds of LivenessProbe: %s", cut,
+						cut.LivenessProbe.PeriodSeconds)
+					compliantObjects = append(compliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID, "Liveness exec probe has a PeriodSeconds greater than 10", true))
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:104]: replace "cut.Satuts.ContainerID" with "cut.Name" here and below

File: cnf-certification-test/performance/suite.go [hunk 28]
```diff
@@ -79,6 +90,66 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+	var compliantObjects []*testhelper.ReportObject
+	var nonCompliantObjects []*testhelper.ReportObject
+	counter := 0
+	for _, put := range env.Pods {
+		for _, cut := range put.Containers {
+			if cut.LivenessProbe != nil && cut.LivenessProbe.Exec != nil {
+				counter++
+				if cut.LivenessProbe.PeriodSeconds > minExecProbePeriodSeconds {
+					tnf.ClaimFilePrintf("Container %s is using exec probes, PeriodSeconds of LivenessProbe: %s", cut,
+						cut.LivenessProbe.PeriodSeconds)
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:103]: Add the actual exec probe period in the report objects: 
compliantObjects = append(compliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Name, fmt.Sprintf("Liveness exec probe has a PeriodSeconds greater than 10 ( %d seconds)", cut.LivenessProbe.PeriodSeconds), true))

File: cnf-certification-test/performance/suite.go (hunk 26 above)
Comment (edcdavid) [reply cnf-certification-test/performance/suite.go:141]: Also update the report objects like: 
nonCompliantObjects = append(nonCompliantObjects, testhelper.NewReportObject(fmt.Sprintf("CNF has 10 or more exec probes (%d exec probes)",counter), testhelper.CnfType, false).

File: cnf-certification-test/performance/suite.go [hunk 29]
```diff
@@ -116,15 +109,42 @@ func testLimitedUseOfExecProbes(env *provider.TestEnvironment) {
...
 				counter++
-				compliantObjects, nonCompliantObjects = CheckProbePeriodSeconds(cut.ReadinessProbe, compliantObjects, nonCompliantObjects, put, cut, "ReadinessProbe")
+				if CheckProbePeriodSeconds(cut.ReadinessProbe, cut, "ReadinessProbe") {
+					compliantObjects = append(compliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name,
+						cut.Name, fmt.Sprintf("ReadinessProbe exec probe has a PeriodSeconds greater than 10 ( %d seconds)",
+							cut.StartupProbe.PeriodSeconds), true))
+				} else {
+					nonCompliantObjects = append(nonCompliantObjects,
+						testhelper.NewContainerReportObject(put.Namespace, put.Name,
+							cut.Status.ContainerID, fmt.Sprintf("ReadinessProbe exec probe has a PeriodSeconds that is not greater than 10 ( %d seconds)",
```
Comment (edcdavid) [inline cnf-certification-test/performance/suite.go:145]: There are some left over cut.Status.ContainerID. Change to cut.name across PR

**Review by edcdavid — APPROVED** [review]
LGTM pending Brandon's comments

File: cnf-certification-test/performance/suite.go [hunk 30]
```diff
@@ -79,6 +91,77 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
 	// Scheduling related tests ends here
 })
 
+func CheckProbePeriodSeconds(elem *v1.Probe, cut *provider.Container, s string) bool {
```
Comment (sebrandon1) [inline cnf-certification-test/performance/suite.go:94]: I feel like you could remove `s` and determine the type of the probe with `reflect` to not have to pass in the string of what type of probe we're looking at.

Comment (shimritproj) [reply cnf-certification-test/performance/suite.go:94]: I was trying to find field that give this data(if is StartupProbe/ LivenessProbe / ReadinessProbe)  but i didn't find.

**Review by sebrandon1 — APPROVED** [review]

File: cnf-certification-test/performance/suite.go [hunk 31]
```diff
@@ -79,6 +90,69 @@ var _ = ginkgo.Describe(common.PerformanceTestKey, func() {
...
+						cut.ReadinessProbe.PeriodSeconds)
+					compliantObjects = append(compliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID, "Ensure limited is not using of exec probes", true))
+				} else {
+					tnf.ClaimFilePrintf("Container %s is not using of exec probes, PeriodSeconds of ReadinessProbe: %s", cut,
+						cut.ReadinessProbe.PeriodSeconds)
+					nonCompliantObjects = append(nonCompliantObjects, testhelper.NewContainerReportObject(put.Namespace, put.Name, cut.Status.ContainerID,
+						"Ensure limited is use of exec probes", false))
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (edcdavid) [reply cnf-certification-test/performance/suite.go]: Why not do something similar to the following:
```func testProbe(aProbe *v1.Probe, counter *int, cut *provider.Container) (nonCompliantObjects, compliantObjects []*testhelper.ReportObject) {
	if aProbe != nil && aProbe.Exec != nil {
		*counter++
		if CheckProbePeriodSeconds(aProbe, cut, "LivenessProbe") {
			compliantObjects = append(compliantObjects, testhelper.NewContainerReportObject(cut.Namespace, cut.Podname,
				cut.Name, fmt.Sprintf("LivenessProbe exec probe has a PeriodSeconds greater than 10 ( %d seconds)",
					cut.LivenessProbe.PeriodSeconds), true))
		} else {
			nonCompliantObjects = append(nonCompliantObjects,
				testhelper.NewContainerReportObject(cut.Namespace, cut.Podname,
					cut.Name, fmt.Sprintf("LivenessProbe exec probe has a PeriodSeconds that is not greater than 10 ( %d seconds)",
						aProbe.PeriodSeconds), false))
		}
	}
	return nonCompliantObjects, compliantObjects
}

func testLimitedUseOfExecProbes(env *provider.TestEnvironment) {
	var compliantObjects []*testhelper.ReportObject
	var nonCompliantObjects []*testhelper.ReportObject
	counter := 0
	for _, put := range env.Pods {
		for _, cut := range put.Containers {
			nonCompliantObjects, compliantObjects := testProbe(cut.LivenessProbe, &counter, cut)

			nonCompliant, compliant := testProbe(cut.StartupProbe, &counter, cut)
			compliantObjects = append(compliantObjects, compliant...)
			nonCompliantObjects = append(nonCompliantObjects, nonCompliant...)

			nonCompliant, compliant = testProbe(cut.ReadinessProbe, &counter, cut)
			compliantObjects = append(compliantObjects, compliant...)
			nonCompliantObjects = append(nonCompliantObjects, nonCompliant...)
		}
	}

	// If there >=10 exec probes, mark the entire cluster as a failure
	if counter >= maxNumberOfExecProbes {
		tnf.ClaimFilePrintf(fmt.Sprintf("CNF has %d exec probes", counter))
		nonCompliantObjects = append(nonCompliantObjects, testhelper.NewReportObject(fmt.Sprintf("CNF has 10 or more exec probes (%d exec probes)", counter), testhelper.CnfType, false))
	} else {
		// Compliant object
		compliantObjects = append(compliantObjects, testhelper.NewReportObject(fmt.Sprintf("CNF has less than 10 exec probes (%d exec probes)", counter), testhelper.CnfType, true))
		tnf.ClaimFilePrintf(fmt.Sprintf("CNF has less than %d exec probes", counter))
	}

	testhelper.AddTestResultReason(compliantObjects, nonCompliantObjects, tnf.ClaimFilePrintf, ginkgo.Fail)
}`

[the remark this answers is not in the document]
Comment (shimritproj) [reply cnf-certification-test/performance/suite.go]: I created new PR and adding the changes. 
tnx @edcdavid  for good comment !

**Pull request merged**